PWA-2915: Added Tailwind CSS configuration and theming documentation (#155)
AdobeDocs/commerce-pwa-studio: src/pages/guides/theming/index.md

---
title: Theming for PWA Studio apps
---

# Theming PWA Studio apps

When creating a new PWA Studio app with [@magento/create-pwa](https://www.npmjs.com/package/@magento/create-pwa), theming is configured out of the box to use Tailwind CSS. The CLI installs all the supporting Tailwind CSS packages and configuration files described below. For a PWA Studio app that was not scaffolded with the CLI, this topic describes how to configure your app for theming with Tailwind CSS.

## Step 1: Install packages

Install the following packages using `npm install` or `yarn add` as follows:

```bash
yarn add -D tailwindcss@~2.2.19 postcss@~8.3.11 postcss-loader@~4.3.0 autoprefixer@latest
```

> 

Author: Done

- **matcher** - The `matcher` variable is used to match CSS Modules classnames that use the `composes` function. This is used by PurgeCSS to remove unused CSS from the `index.css` file. See [Writing purgeable HTML](https://v2.tailwindcss.com/docs/optimizing-for-production#writing-purgeable-html) from the Tailwind documentation.

<InlineAlert slots="text" />

When looking at code samples in the Tailwind docs, keep in mind that the base project configures Tailwind to use `_` as the [seperator](https://v2.tailwindcss.com/docs/configuration#separator) instead of the default colon (`:`).

### Add postcss.config.js

The `postcss.config.js` file is a standard configuration file used by the `postcss-loader` package. So you need to add it to your project's root and configure it to add `tailwindcss` and the `autoprefixer` as a PostCSS plugins for your app. Your `postcss.config.js` file should look like this:

The `create-pwa` CLI adds and configures this file as follows:

```js
// postcss.config.js

module.exports = {
  plugins: [
    require('autoprefixer'),
    require('tailwindcss')('./tailwind.config.js'),
  ],
};
```

## Step 3: Test your configuration

> To test your configuration with the `[venia]` theme set as a `preset` in your config file, add the following `theme` style override and rebuild your app.

Reviewer: At the beginning of this topic, you described that it was meant for those who didn't use the default scaffolding. I was willing to give the benefit of the doubt that step 2 was just for setting up in an alternate environment, but really, it does contain advice for changing the defaults. As of this step though, it's clear that it applies to everyone. 

Rewrite the intro and make sure it's clear that 

* Step 1 provides details about the two installation paths
* Alternate scaffolding users must go through step 2, but the happy path installers might need to make changes from the default configuration
* Steps 3 and 4 are for everyone.

Reviewer: Below, in option 2, you say modifying the `config.js` file isn't the right thing to do in most cases. I recommend saying that this step is a quick test, and you might want to back out these changes when creating your own presets.

```js
// tailwind.config.js

theme: {
  extend: {
    textColor: {
      colorDefault: '#BB0000'
    }
  }
}
```

If the text color in your Venia-themed app turns red, your configuration is working correctly. To test your own theme, refer to [Customizing your theme](https://tailwindcss.com/docs/adding-custom-styles#customizing-your-theme) from the Tailwind documentation.

## Step 4: Customize the base styles

You have two options for customizing the Tailwind and Venia base styles:

1. Update the `tailwind.config.js` file directly.
1. Import your own theme file as a `preset` (recommended).

### Option 1: Update tailwind.config.js theme directly

> If you followed Step 3 above, you have already done this as a test. Another example: change the default font family by adding new styles to the `tailwind.config.js` `theme` property:

Reviewer: I would assume that the reader followed step 3 (no need to write "above"). Add the fontFamily example below to step 3, and use this section point out this is what you did, but you probably don't want to use it in a real-world scenario.

1. Create a new `my-theme-preset.js` file anywhere in your project and name it whatever you want.
1. Export a new `theme` object with your Tailwind and/or Venia base theme extensions and overrides:

  ```js
  // my-theme-preset.js

  module.exports = {
    theme: {
      extend: {
        fontFamily: {
          sans: ['Courier', 'sans-serif']
        }
      }
    }
  }
  ```

1. Import your theme preset to the `tailwind.config.js` and add it to the `presets` array:

  ```js
  // tailwind.config.js

  const venia = require('@magento/pwa-theme-venia');
  const myThemePreset = require('./my-theme-preset');

  module.exports = {
    presets: [venia, myThemePreset]
  }
  ```

  This example adds the `myThemePreset` after the `venia` preset, meaning that myThemePreset will be merged and override or extend the base styles from Venia and Tailwind. For more information, see [how configurations are merged](https://tailwindcss.com/docs/presets#merging-logic-in-depth) from the Tailwind documentation.

1. Rebuild your app.

### Override or extend Tailwind base styles

> To override or extend the Tailwind base styles for version `2.2.19`, and any other version, use this site as a [reference to the Tailwind base styles](https://unpkg.com/browse/tailwindcss@2.2.19/stubs/defaultConfig.stub.js).

Reviewer: Can you say something like "the base styles for any supported version of TW"? Again, trying to future-proof the docs.

For example, to **override** Venia's default `fontFamily` options and **extend** its default `backgroundColor` and `borderRadius`, you could add the following `theme` properties to your preset:

```js
// my-theme-preset.js

module.exports = {
  theme: {
    extend: {
      // Override Venia's default font family
      fontFamily: {
        sans: ['Open Source', 'sans-serif'],
        serif: ['Gothic', 'serif']
      },
      // Extend Venia's backgroundColor and borderRadius styles
      backgroundColor: {
        subtitle: '#F5F5F5',
      }
      borderRadius: {
        radius4: '50%',
    }
  },
```

Another example: To change the width of Venia's sidebar filter on the search and category pages, you can override the `theme.extend.spacing.filterSidebarWidth`, as shown here:

```js
// my-theme-preset.js

module.exports = {
  theme: {
    extend: {
      spacing: {
        filterSidebarWidth: '400px'
      }
    }
  }
}
```

## Useful extension examples

The following extension examples show some additional useful customizations you can make to Tailwind and Venia base themes.

> ### Adding new utility classes

Reviewer: ```suggestion
### Add new utility classes
```

As described above with extensions, Tailwind gives you the ability to define new utility classes with your own values. You can use these values in your own custom CSS files for your custom components or style overrides for Venia components.

For example, you can add new [height utilities](https://tailwindcss.com/docs/height) by specifying entries for `theme.extend.height`.

```js
// my-theme-preset.js

module.exports = {
  theme: {
    extend: {
      height: {
        customHeight: '16rem',
      }
    }
  }
}
```

This creates a new height utility `height-customHeight` which you can use in your project.

> ### Replacing existing breakpoint values

Reviewer: ```suggestion
### Replace existing breakpoint values
```

```js:

```js
// tailwind.config.js

theme: {
  extend: {
    textColor: {
      colorDefault: '#BB0000'
    }
  }
}
```

To test your own theme, refer to [Customizing your theme](https://tailwindcss.com/docs/adding-custom-styles#customizing-your-theme) from the Tailwind documentation.

## Customize base styles

You have two options for customizing the Tailwind and Venia base styles:

1. Update the `tailwind.config.js` file directly.
1. Import your own theme file as a `preset` (recommended).

### Update tailwind.config.js directly

As you did to test the theme configuration, you can override or extend the Tailwind and Venia themes directly within the `theme` property of the `tailwind.config.js.` file as shown here:

```js
// tailwind.config.js

theme: {
  extend: {
    fontFamily: {
      sans: ['"Open Sans"', 'sans-serif']
    }
    backgroundColor: {
      subtitle: '#F5F5F5',
    }
    borderRadius: {
      radius4: '50%',
    }
    ...
  }
}
```

> However, you should avoid this option for all but the smallest real-world changes. For example, if you plan to use the Venia base theme with only a few custom colors and fonts, this option may be fine for your needs. However, if you are going to make several style changes or plan to rotate through different themes (seasonal or otherwise), use option 2.

Reviewer: ```suggestion
However, you should avoid this option for all but the smallest real-world changes. For example, if you plan to use the Venia base theme with only a few custom colors and fonts, this option may be fine for your needs. However, if you are going to make several style changes or plan to rotate through different themes (seasonal or otherwise), create a new theme file instead.
```

### Create new theme file

We recommended this option for all but the smallest changes. The steps for creating your own [theme presets](https://tailwindcss.com/docs/presets) and adding them to the `tailwind.config.js` file to customize the existing base themes are as follows:

1. Create a new `my-theme-preset.js` file anywhere in your project and name it whatever you want.
1. Export a new `theme` object with your Tailwind and/or Venia base theme extensions and overrides:

  ```js
  // my-theme-preset.js

  module.exports = {
    theme: {
      extend: {
        fontFamily: {
          sans: ['Courier', 'sans-serif']
        }
      }
    }
  }
  ```

1. Import your theme preset to the `tailwind.config.js` and add it to the `presets` array:

  ```js
  // tailwind.config.js

  const venia = require('@magento/pwa-theme-venia');
  const myThemePreset = require('./my-theme-preset');

  module.exports = {
    presets: [venia, myThemePreset]
  }
  ```

>   This example adds the `myThemePreset` after the `venia` preset, meaning that myThemePreset will be merged and override or extend the base styles from Venia and Tailwind. For more information, see [how configurations are merged](https://tailwindcss.com/docs/presets#merging-logic-in-depth) from the Tailwind documentation.

Reviewer: ```suggestion
  This example adds `myThemePreset` after the `venia` preset, meaning that `myThemePreset` will be merged and override or extend the base styles from Venia and Tailwind. For more information, see [how configurations are merged](https://tailwindcss.com/docs/presets#merging-logic-in-depth) from the Tailwind documentation.
```

      backgroundColor: {
        subtitle: '#F5F5F5',
      }
      borderRadius: {
        radius4: '50%',
    }
  },
```

Another example: To change the width of Venia's sidebar filter on the search and category pages, you can override the `theme.extend.spacing.filterSidebarWidth`, as shown here:

```js
// my-theme-preset.js

module.exports = {
  theme: {
    extend: {
      spacing: {
        filterSidebarWidth: '400px'
      }
    }
  }
}
```

## Useful examples

The following extension examples show some additional useful customizations you can make to Tailwind and Venia base themes.

### Add new utility classes

As described above with extensions, Tailwind gives you the ability to define new utility classes with your own values. You can use these values in your own custom CSS files for your custom components or style overrides for Venia components.

For example, you can add new [height utilities](https://tailwindcss.com/docs/height) by specifying entries for `theme.extend.height`.

```js
// my-theme-preset.js

module.exports = {
  theme: {
    extend: {
      height: {
        customHeight: '16rem',
      }
    }
  }
}
```

> This creates a new height utility `height-customHeight` which you can use in your project.

Reviewer: ```suggestion
This creates a new height utility `height-customHeight`, which you can use in your project.
```

---
title: Tailwind Theming
description: Learn how to use Tailwind CSS to theme your PWA Studio app.
featured: true
keywords: [Theming, Tailwind CSS, CSS variables, PWA Studio]
---

# Tailwind Theming for PWA Studio apps

[Tailwind CSS](https://tailwindcss.com/) is a popular CSS framework for rapidly building custom user interfaces, focused on flexibility and designed for quick, custom branding. This topic describes how you can install, configure, and use Tailwind with Venia to theme your own PWA Studio app.

## Tailwind and Venia themes

> Both the Tailwind and Venia themes are considered base themes — themes designed with only minimal brand-styling. We use base themes so you can quickly customize your apps with different colors, typography, images, and other brand attributes. The included Venia base theme provides a layer on top of Tailwind's base theme. The Venia theme uses all the best practices required for overriding and extending other base themes.

Reviewer 2: ```suggestion
Both the Tailwind and Venia themes are considered base themes — themes designed with only minimal brand-styling. We use base themes so you can quickly customize your apps with different colors, typography, images, and other brand attributes. The included [Venia base theme](https://github.com/magento/pwa-studio/tree/develop/packages/pwa-theme-venia) provides a layer on top of Tailwind's base theme. The Venia theme uses all the best practices required for overriding and extending other base themes.
```
